File: iOS/swiftui-mfa/MfaSample/MfaSample/Core/Managers/AccountGroupingManager.swift
//
//  AccountGroupingManager.swift
//  MfaSample
//
//  Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
//
//  This software may be modified and distributed under the terms
//  of the MIT license. See the LICENSE file for details.
//

import Foundation
import PingOath
import PingPush
import Combine

/// Manager for grouping and organizing credentials into AccountGroups.
@MainActor
class AccountGroupingManager: ObservableObject {
    static let shared = AccountGroupingManager()

    private let userPreferences = UserPreferences.shared

    private init() {}

    /// Groups credentials by issuer and account name based on user preferences.
    /// - Parameters:
    ///   - oathCredentials: List of OATH credentials to group.
    ///   - pushCredentials: List of Push credentials to group.
    ///   - shouldCombine: Whether to combine credentials with the same issuer and account name.
    /// - Returns: Array of AccountGroups.
    func groupCredentialsByAccount(
        oathCredentials: [OathCredential],
        pushCredentials: [PushCredential],
        shouldCombine: Bool? = nil
    ) -> [AccountGroup] {
        let combine = shouldCombine ?? userPreferences.combineAccounts

        if !combine {
            // Create separate account groups for each credential
            return createSeparateGroups(oathCredentials: oathCredentials, pushCredentials: pushCredentials)
        }

        // Combine accounts with the same issuer and account name
        return createCombinedGroups(oathCredentials: oathCredentials, pushCredentials: pushCredentials)
    }

    /// Creates separate AccountGroups for each credential (no combining).
    private func createSeparateGroups(
        oathCredentials: [OathCredential],
        pushCredentials: [PushCredential]
    ) -> [AccountGroup] {
        var groups: [AccountGroup] = []

        // Create individual OATH account groups
        for credential in oathCredentials {
            groups.append(AccountGroup(
                issuer: credential.issuer,
                accountName: credential.accountName,
                displayIssuer: credential.displayIssuer,
                displayAccountName: credential.displayAccountName,
                oathCredentials: [credential],
                pushCredentials: []
            ))
        }

        // Create individual Push account groups
        for credential in pushCredentials {
            groups.append(AccountGroup(
                issuer: credential.issuer,
                accountName: credential.accountName,
                displayIssuer: credential.displayIssuer,
                displayAccountName: credential.displayAccountName,
                oathCredentials: [],
                pushCredentials: [credential]
            ))
        }

        return groups
    }

    /// Creates combined AccountGroups where credentials with the same issuer and account name are grouped together.
    private func createCombinedGroups(
        oathCredentials: [OathCredential],
        pushCredentials: [PushCredential]
    ) -> [AccountGroup] {
        var accountGroups: [String: AccountGroup] = [:]

        // Add OATH credentials to groups
        for credential in oathCredentials {
            let key = "\(credential.issuer)-\(credential.accountName)"
            if let existingGroup = accountGroups[key] {
                // Update existing group
                accountGroups[key] = AccountGroup(
                    issuer: existingGroup.issuer,
                    accountName: existingGroup.accountName,
                    displayIssuer: credential.displayIssuer,
                    displayAccountName: credential.displayAccountName,
                    oathCredentials: existingGroup.oathCredentials + [credential],
                    pushCredentials: existingGroup.pushCredentials
                )
            } else {
                // Create new group
                accountGroups[key] = AccountGroup(
                    issuer: credential.issuer,
                    accountName: credential.accountName,
                    displayIssuer: credential.displayIssuer,
                    displayAccountName: credential.displayAccountName,
                    oathCredentials: [credential],
                    pushCredentials: []
                )
            }
        }

        // Add Push credentials to groups
        for credential in pushCredentials {
            let key = "\(credential.issuer)-\(credential.accountName)"
            if let existingGroup = accountGroups[key] {
                // Update existing group
                accountGroups[key] = AccountGroup(
                    issuer: existingGroup.issuer,
                    accountName: existingGroup.accountName,
                    displayIssuer: credential.displayIssuer,
                    displayAccountName: credential.displayAccountName,
                    oathCredentials: existingGroup.oathCredentials,
                    pushCredentials: existingGroup.pushCredentials + [credential]
                )
            } else {
                // Create new group
                accountGroups[key] = AccountGroup(
                    issuer: credential.issuer,
                    accountName: credential.accountName,
                    displayIssuer: credential.displayIssuer,
                    displayAccountName: credential.displayAccountName,
                    oathCredentials: [],
                    pushCredentials: [credential]
                )
            }
        }

        return Array(accountGroups.values)
    }

    /// Applies saved ordering to account groups.
    /// - Parameter groups: The account groups to order.
    /// - Returns: Ordered account groups.
    func applyOrdering(to groups: [AccountGroup]) -> [AccountGroup] {
        let savedOrder = userPreferences.getAccountOrder()

        if savedOrder.isEmpty {
            return groups
        }

        // Check if we have separate groups (when combine accounts is disabled)
        let hasSeparateGroups = groups.contains { group in
            groups.filter { $0.issuer == group.issuer && $0.accountName == group.accountName }.count > 1
        }

        return hasSeparateGroups
            ? applySeparateGroupOrdering(groups, savedOrder: savedOrder)
            : applyCombinedGroupOrdering(groups, savedOrder: savedOrder)
    }

    /// Applies ordering for separate account groups (when combine accounts is disabled).
    /// Groups OATH and Push cards from the same account together and maintains saved order.
    private func applySeparateGroupOrdering(_ accountGroups: [AccountGroup], savedOrder: [String]) -> [AccountGroup] {
        // Group the separate cards by account (issuer + accountName)
        let groupsByAccount = Dictionary(grouping: accountGroups) { "\($0.issuer)-\($0.accountName)" }

        var orderedList: [AccountGroup] = []
        var processedAccounts = Set<String>()

        // Process accounts in saved order
        for accountKey in savedOrder {
            if let accountCards = groupsByAccount[accountKey], !processedAccounts.contains(accountKey) {
                // Sort cards within account: OATH first, then Push
                let sortedCards = accountCards.sorted { a, b in
                    if !a.oathCredentials.isEmpty && !b.pushCredentials.isEmpty {
                        return true  // OATH first
                    } else if !a.pushCredentials.isEmpty && !b.oathCredentials.isEmpty {
                        return false  // Push second
                    }
                    return false  // Same type, maintain existing order
                }
                orderedList.append(contentsOf: sortedCards)
                processedAccounts.insert(accountKey)
            }
        }

        // Add any new accounts not in saved order
        for (accountKey, accountCards) in groupsByAccount {
            if !processedAccounts.contains(accountKey) {
                let sortedCards = accountCards.sorted { a, b in
                    if !a.oathCredentials.isEmpty && !b.pushCredentials.isEmpty {
                        return true
                    } else if !a.pushCredentials.isEmpty && !b.oathCredentials.isEmpty {
                        return false
                    }
                    return false
                }
                orderedList.append(contentsOf: sortedCards)
            }
        }

        return orderedList
    }

    /// Applies ordering for combined account groups (when combine accounts is enabled).
    private func applyCombinedGroupOrdering(_ accountGroups: [AccountGroup], savedOrder: [String]) -> [AccountGroup] {
        // Create a map for quick lookup
        let accountMap = Dictionary(uniqueKeysWithValues: accountGroups.map { ("\($0.issuer)-\($0.accountName)", $0) })

        var orderedList: [AccountGroup] = []
        var addedKeys = Set<String>()

        // Add accounts in saved order
        for key in savedOrder {
            if let accountGroup = accountMap[key], !addedKeys.contains(key) {
                orderedList.append(accountGroup)
                addedKeys.insert(key)
            }
        }

        // Add any new accounts that weren't in the saved order
        for accountGroup in accountGroups {
            let key = "\(accountGroup.issuer)-\(accountGroup.accountName)"
            if !addedKeys.contains(key) {
                orderedList.append(accountGroup)
            }
        }

        return orderedList
    }

    /// Saves the current order of account groups.
    /// - Parameter groups: The ordered account groups.
    func saveOrdering(for groups: [AccountGroup]) {
        let order = groups.map { "\($0.issuer)-\($0.accountName)" }
        userPreferences.setAccountOrder(order)
    }
}

File: android/kotlin-authenticatorapp/app/src/main/kotlin/com/pingidentity/authenticatorapp/managers/AccountGroupingManager.kt
/*
 * Copyright (c) 2025 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package com.pingidentity.authenticatorapp.managers

import com.pingidentity.authenticatorapp.data.AccountGroup
import com.pingidentity.authenticatorapp.data.DiagnosticLogger
import com.pingidentity.authenticatorapp.data.UserPreferences
import com.pingidentity.authenticatorapp.data.groupCredentialsByAccount
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import com.pingidentity.mfa.oath.OathCredential
import com.pingidentity.mfa.push.PushCredential

/**
 * Manager class for handling account grouping logic and account ordering.
 * Encapsulates the business logic for combining and organizing account groups.
 *
 * @param userPreferences UserPreferences dependency for settings management
 * @param diagnosticLogger DiagnosticLogger for logging
 */
class AccountGroupingManager(
    private val userPreferences: UserPreferences,
    private val diagnosticLogger: DiagnosticLogger
) {
    
    private val _accountGroups = MutableStateFlow<List<AccountGroup>>(emptyList())
    val accountGroups: StateFlow<List<AccountGroup>> = _accountGroups.asStateFlow()

    /**
     * Updates the account groups based on current credentials and user preferences.
     */
    fun updateAccountGroups(
        oathCredentials: List<OathCredential>,
        pushCredentials: List<PushCredential>
    ) {
        val shouldCombine = userPreferences.isCombineAccountsEnabled()
        diagnosticLogger.d("updateAccountGroups: shouldCombine=$shouldCombine")
        
        // Skip update if we have no data to avoid unnecessary recomposition
        if (oathCredentials.isEmpty() && pushCredentials.isEmpty()) {
            diagnosticLogger.d("Skipping account group update - no credentials loaded yet")
            return
        }

        val newAccountGroups = groupCredentialsByAccount(
            oathCredentials, pushCredentials, shouldCombine
        )
        
        // Apply saved account order
        val orderedAccountGroups = applyAccountOrder(newAccountGroups)
        
        // Debug logging to help identify duplicate group issues
        diagnosticLogger.d("updateAccountGroups: shouldCombine=$shouldCombine, " +
                "oathCredentials=${oathCredentials.size}, " +
                "pushCredentials=${pushCredentials.size}, " +
                "resultingGroups=${orderedAccountGroups.size}")
        
        // Validate that we don't have duplicate account groups with the same issuer+account name
        val groupKeys = orderedAccountGroups.map { "${it.issuer}-${it.accountName}" }
        val duplicateKeys = groupKeys.groupBy { it }.filter { it.value.size > 1 }.keys
        if (duplicateKeys.isNotEmpty()) {
            diagnosticLogger.w("Warning: Found duplicate account groups with keys: $duplicateKeys")
            // This is expected when shouldCombine is false and there are multiple credentials
            // for the same account, but worth logging for debugging
        }

        _accountGroups.value = orderedAccountGroups
    }

    /**
     * Updates the account group order immediately.
     * This provides immediate feedback while the order is being persisted.
     */
    fun updateAccountGroupOrder(newAccountGroups: List<AccountGroup>) {
        diagnosticLogger.d("Update AccountGroupOrder")
        _accountGroups.value = newAccountGroups
    }

    /**
     * Save the current account order to preferences.
     */
    suspend fun saveAccountOrder(accountGroups: List<AccountGroup>) {
        val orderKeys = accountGroups.map { "${it.issuer}-${it.accountName}" }
        userPreferences.setAccountOrder(orderKeys)
    }

    /**
     * Apply saved account order to the list of account groups.
     */
    private fun applyAccountOrder(accountGroups: List<AccountGroup>): List<AccountGroup> {
        val savedOrder = userPreferences.getAccountOrder()
        if (savedOrder.isEmpty()) {
            return accountGroups
        }

        // Check if we have separate groups (when combine accounts is disabled)
        val hasSeparateGroups = accountGroups.any { group ->
            accountGroups.count { it.issuer == group.issuer && it.accountName == group.accountName } > 1
        }

        return if (hasSeparateGroups) {
            applySeparateGroupOrdering(accountGroups, savedOrder)
        } else {
            applyCombinedGroupOrdering(accountGroups, savedOrder)
        }
    }

    /**
     * Apply ordering for separate account groups (when combine accounts is disabled).
     * Groups OATH and Push cards from the same account together and maintains saved order.
     */
    private fun applySeparateGroupOrdering(
        accountGroups: List<AccountGroup>,
        savedOrder: List<String>
    ): List<AccountGroup> {
        // Group the separate cards by account (issuer + accountName)
        val groupsByAccount = accountGroups.groupBy { "${it.issuer}-${it.accountName}" }

        val orderedList = mutableListOf<AccountGroup>()
        val processedAccounts = mutableSetOf<String>()

        // Process accounts in saved order
        savedOrder.forEach { accountKey ->
            groupsByAccount[accountKey]?.let { accountCards ->
                // Sort cards within account: OATH first, then Push
                val sortedCards = accountCards.sortedWith { a, b ->
                    when {
                        a.oathCredentials.isNotEmpty() && b.pushCredentials.isNotEmpty() -> -1 // OATH first
                        a.pushCredentials.isNotEmpty() && b.oathCredentials.isNotEmpty() -> 1   // Push second
                        else -> 0 // Same type, maintain existing order
                    }
                }
                orderedList.addAll(sortedCards)
                processedAccounts.add(accountKey)
            }
        }

        // Add any new accounts not in saved order
        groupsByAccount.entries.forEach { (accountKey, accountCards) ->
            if (!processedAccounts.contains(accountKey)) {
                val sortedCards = accountCards.sortedWith { a, b ->
                    when {
                        a.oathCredentials.isNotEmpty() && b.pushCredentials.isNotEmpty() -> -1
                        a.pushCredentials.isNotEmpty() && b.oathCredentials.isNotEmpty() -> 1
                        else -> 0
                    }
                }
                orderedList.addAll(sortedCards)
            }
        }

        return orderedList
    }

    /**
     * Apply ordering for combined account groups (when combine accounts is enabled).
     */
    private fun applyCombinedGroupOrdering(
        accountGroups: List<AccountGroup>,
        savedOrder: List<String>
    ): List<AccountGroup> {
        // Create a map for quick lookup (only for combined accounts)
        val accountMap = accountGroups.associateBy { "${it.issuer}-${it.accountName}" }
        val orderedList = mutableListOf<AccountGroup>()
        val addedKeys = mutableSetOf<String>()

        // Add accounts in saved order
        savedOrder.forEach { key ->
            accountMap[key]?.let { accountGroup ->
                orderedList.add(accountGroup)
                addedKeys.add(key)
            }
        }

        // Add any new accounts that weren't in the saved order
        accountGroups.forEach { accountGroup ->
            val key = "${accountGroup.issuer}-${accountGroup.accountName}"
            if (!addedKeys.contains(key)) {
                orderedList.add(accountGroup)
            }
        }

        return orderedList
    }
}